chaindataloader/binary.go

```diff
@@ -0,0 +1,227 @@
+	chain        wire.BitcoinNet
+	dataTypeSize uint32
+}
+
+// newBinaryChainDataLoader initializes a Binary Reader.
+func newBinaryChainDataLoader(reader io.ReadSeeker,
+	typeData dataType) (*binReader, error) {
+
+	// Read data type.
+	dataTypeBytes := [8]byte{}
+	typeOfData, err := tlv.ReadVarInt(reader, &dataTypeBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining data type "+
+			"of file %w", err)
+	}
+
+	if dataType(typeOfData) != typeData {
+		return nil, fmt.Errorf("data type mismatch: got %v but expected %v",
+			dataType(typeOfData), typeData)
+	}
+
+	var dataTypeSize uint32
+	switch {
+	case typeData == blockHeaders:
+		dataTypeSize = headerfs.BlockHeaderSize
+	default:
+		return nil, fmt.Errorf("got unsupported dataType: %v", typeData)
+	}
+
+	// Read start height of block header file.
+	startBytes := [8]byte{}
+	start, err := tlv.ReadVarInt(reader, &startBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining start height "+
+			"of file %w", err)
+	}
+
+	// Read end height of block header file.
+	endBytes := [8]byte{}
+	end, err := tlv.ReadVarInt(reader, &endBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining end height "+
+			"of file %w", err)
+	}
+
+	// Read the bitcoin network, the headers in the header file belong to.
+	netBytes := [8]byte{}
+	chainChar, err := tlv.ReadVarInt(reader, &netBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining chain "+
+			"of file %w", err)
+	}
+
+	var chain wire.BitcoinNet
+	switch {
+	case chainChar == 1:
+		chain = wire.TestNet3
+	case chainChar == 2:
+		chain = wire.MainNet
+	case chainChar == 3:
+		chain = wire.SimNet
+	case chainChar == 4:
+		chain = wire.TestNet
+	default:
+		return nil, fmt.Errorf("read unsupported character (%d) for "+
+			"network of side-load source file", chainChar)
+	}
+
+	// obtain space occupied by metadata as initial offset
+	initialOffset, err := reader.Seek(0, io.SeekCurrent)
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to determine initial offset: "+
+			"%v", err)
+	}
+
+	return &binReader{
+		reader:        reader,
+		startHeight:   uint32(start),
+		endHeight:     uint32(end),
+		chain:         chain,
+		initialOffset: uint32(initialOffset),
+		dataTypeSize:  dataTypeSize,
+	}, nil
+}
+
+// NextBlockHeaders fetches headers from the reader. An empty slice is returned
+// if there are no headers left to fetch.
+func (b *binReader) NextBlockHeaders(n uint32) ([]*wire.BlockHeader, error) {
+	headers := make([]*wire.BlockHeader, n)
+
+	createData := func(headers []*wire.BlockHeader) func(data []byte,
+		i uint32) error {
+
+		return func(data []byte, i uint32) error {
+			var blockHeader wire.BlockHeader
+
+			headerReader := bytes.NewReader(data)
+
+			// Finally, decode the raw bytes into a proper bitcoin header.
+			if err := blockHeader.Deserialize(headerReader); err != nil {
+				return fmt.Errorf("error deserializing block header: %w",
+					err)
+			}
+
+			headers[i] = &blockHeader
```

Comment:
It's odd behavior to mutate an argument passed into a closure function. Why doesn't this just instead return the block header directly?

Reply:
Then this also doesn't need to be a higher order function. Instead it just decodes the header, and returns it.